sw/device/lib/testing/test_framework/ujson_ottf.h

```diff
@@ -20,6 +21,34 @@ extern "C" {
  */
 ujson_t ujson_ottf_console(void);
 
+/**
+ * Deserialize a ujson message with a CRC.
+ * This macro will deserialize the message and then deserialize the CRC and
+ * check that it matches the CRC of the message. The CRC is itself deserialize
+ * as the ujson structure `ottf_crc_t`.
+ *
+ * @param deser_fn_ A pointer to the deserialization function generated by
+ * ujson_derive.
+ * @param uj_ctx_ A `ujson_t` representing the IO context.
+ * @param data_ A pointer to the data to send.
+ *
+ */
+#define UJSON_WITH_CRC(deser_fn_, uj_ctx_, data_)                    \
+  ({                                                                 \
+    ujson_crc32_reset(uj_ctx_);                                      \
+    status_t err = deser_fn_(uj_ctx_, data_);                        \
+    uint32_t msg_crc = ujson_crc32_finish(uj_ctx_);                  \
+    ottf_crc_t crc;                                                  \
+    if (status_ok(err)) {                                            \
+      err = ujson_deserialize_ottf_crc_t(uj_ctx_, &crc);             \
+      if (status_ok(err) && msg_crc != crc.crc) {                    \
+        LOG_ERROR("ujson CRC mismatch: %x vs %x", msg_crc, crc.crc); \
+        err = DATA_LOSS();                                           \
+      }                                                              \
+    }                                                                \
+    err;                                                             \
+  })
+
 /**
  * Add a CRC to the response.
  * Should not be used directly.
```

Comment:
It would be nice to use the `OttfCrc` struct for device to host as well, i.e. in `RESP_CRC`, but it would likely be an unnecessary complication. *See https://github.com/lowRISC/opentitan/issues/17496#issuecomment-1523732105*